network-libp2p/src/network.rs

```diff
@@ -1298,6 +1352,149 @@ impl Network {
+                std::any::type_name::<Req>(),
+                Req::MAX_REQUESTS,
+                Req::TIME_WINDOW,
+            );
+            return false;
+        }
+        true
+    }
+
+    fn remove_rate_limits(
+        peer_request_limits: Arc<Mutex<HashMap<PeerId, HashMap<u16, RateLimit>>>>,
+        rate_limits_pending_deletion: Arc<Mutex<VecDeque<((PeerId, u16), TokioInstant)>>>,
+        peer_id: PeerId,
+    ) {
+        // Every time a peer disconnects, we delete all expired pending limits.
+        Self::clean_up(
+            Arc::clone(&peer_request_limits),
+            Arc::clone(&rate_limits_pending_deletion),
+        );
+
+        // Firstly we must acquire the lock of the pending deletes to avoid deadlocks.
+        let mut rate_limits_pending_deletion_l = rate_limits_pending_deletion.lock();
+        let mut peer_request_limits_l = peer_request_limits.lock();
+
+        // Go through all existing request types of the given peer and deletes the limit counters if possible or marks it for deletion.
+        if let Some(request_limits) = peer_request_limits_l.get_mut(&peer_id) {
+            request_limits.retain(|req_type, rate_limit| {
+                // Gets the requests limit and deletes it if no counter info would be lost, otherwise places it as pending deletion.
+                if !rate_limit.can_delete(TokioInstant::now()) {
+                    rate_limits_pending_deletion_l
+                        .push_back(((peer_id, *req_type), rate_limit.next_reset_time()));
+                    true
+                } else {
+                    false
+                }
+            });
+            // If the peer no longer has any pending rate limits, then it gets removed.
+            if request_limits.is_empty() {
+                peer_request_limits_l.remove(&peer_id);
+            }
+        }
+    }
+
+    /// Deletes the rate limits that were previously marked as pending if their expiration block height has passed.
+    fn clean_up(
+        peer_request_limits: Arc<Mutex<HashMap<PeerId, HashMap<u16, RateLimit>>>>,
+        rate_limits_pending_deletion: Arc<Mutex<VecDeque<((PeerId, u16), TokioInstant)>>>,
+    ) {
+        let mut rate_limits_pending_deletion_l = rate_limits_pending_deletion.lock();
+
+        // Iterates from the oldest to the most recent object of the queue, deletes the entries that have expired.
+        // The queue is ordered from the oldest to the most recent object added, meaning that its loosely ordered by expiration date.
+        // Thus, the iteration stops upon the first object that is hasn't yet expired. In the case where the queue is out of order,
+        // the clean up will reach it after a few more blocks (max block_range).
+        while let Some(((peer_id, req_type), block_expiration)) =
+            rate_limits_pending_deletion_l.front()
+        {
+            let current_timestamp = TokioInstant::now();
+            if block_expiration <= &current_timestamp {
```

Comment:
Is it still correct to call it `block_expiration` as it's not based on blocks anymore?

Reply:
nope. Good catch